src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorUpdaterStep.cs

```diff
@@ -0,0 +1,148 @@
+            WellKnownStepIds.SetTFMStepId,
+        };
+
+        public override IEnumerable<string> DependencyOf { get; } = new[]
+        {
+            WellKnownStepIds.NextProjectStepId,
+        };
+
+        private ImmutableArray<RazorCodeDocument>? _razorDocuments;
+
+        public ImmutableArray<RazorCodeDocument> RazorDocuments =>
+            _razorDocuments ?? throw new InvalidOperationException("Razor documents are not available until the updater step is initialized or until ProcessRazorDocuments has been called.");
+
+        public RazorUpdaterStep(IEnumerable<IUpdater<RazorCodeDocument>> razorUpdaters, ILogger<RazorUpdaterStep> logger)
+            : base(logger)
+        {
+            if (razorUpdaters is null)
+            {
+                throw new ArgumentNullException(nameof(razorUpdaters));
+            }
+
+            if (logger is null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            // Add a sub-step for each Razor updater
+            SubSteps = new List<UpgradeStep>(razorUpdaters.Select(updater => new RazorUpdaterSubStep(this, updater, logger)));
+        }
+
+        /// <summary>
+        /// Determines whether the RazorUpdaterStep applies to a given context. This doesn't
+        /// consider whether there's actually any work to be done by Razor updaters. Rather, it
+        /// determines whether it makes sense to even initialize the Razor updaters. So, it
+        /// will return false if there are no cshtml files in the current project or if there
+        /// are no Razor updaters available.
+        /// </summary>
+        /// <param name="context">The context to evaluate.</param>
+        /// <param name="token">A token that can be used to cancel execution.</param>
+        /// <returns>True if the Razor updater step might apply, false otherwise.</returns>
+        protected override Task<bool> IsApplicableImplAsync(IUpgradeContext context, CancellationToken token) =>
+            Task.FromResult(context?.CurrentProject is not null
+            && SubSteps.Any()
+            && GetRazorFileSystem(context.CurrentProject.Required()).EnumerateItems("/").Any());
+
+        protected override async Task<UpgradeStepInitializeResult> InitializeImplAsync(IUpgradeContext context, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            _razorEngine = RazorProjectEngine.Create(RazorConfiguration.Default, GetRazorFileSystem(context.CurrentProject.Required()), builder =>
+            {
+                // Add useful "MVC View-isms" like _ViewImports support, page models, etc.
+                // https://github.com/dotnet/aspnetcore/blob/main/src/Razor/Microsoft.AspNetCore.Mvc.Razor.Extensions/src/RazorExtensions.cs
+                RazorExtensions.Register(builder);
+            });
+
+            ProcessRazorDocuments();
```

Comment:
I think this says we have to process razor documents foreach substep which would be a perf hit. I can't think of a reasonable workaround but it seems like something we should be fully aware of before adding.

Reply:
Yes, I was thinking about this too. I think a better design will be for sub-steps to indicate which files they changed (if any) and only re-process those documents. I'll see what that looks like.

Reply:
I've fixed this. Updaters now return a list of files they changed and the sub-step only calls for those to be re-processed.